web/ui/react-app/src/PanelList.tsx

```diff
@@ -88,20 +94,24 @@ class PanelList extends Component<any, PanelListState> {
     this.setMetrics();
   }
 
-  setMetrics = () => {
+  setMetrics = (metrics?: string[]) => {
+    const { metricNames, queries } = this.state.metricGroups;
+    let queryItems: string[] = [];
     if (this.isHistoryEnabled()) {
       const historyItems = this.getHistoryItems();
       const { length } = historyItems;
-      this.setState({
-        metricNames: [...historyItems.slice(length - 50, length), ...this.initialMetricNames],
-      });
-    } else {
-      this.setState({ metricNames: this.initialMetricNames });
+      queryItems = historyItems.slice(length - 50, length)
```

Comment:
Currently we don't seem to be keeping a correct rolling window of the past N (e.g. 50) queries here and in the local storage. The local storage in `handleQueryHistory()` currently keeps everything without limit, but at the same time, the autocomplete doesn't keep the last 50 correctly somehow.

Simple example: Try executing as queries the numbers 1 ... 60. You'll see all 60 of them in the local storage, but if you type `6`, you will not see an autocomplete entry for `60`.

I think what we want:
- always limit the local storage to the last N queries
- of course still don't store or show duplicate queries twice (currently correct)
- if a past query gets executed again, update it to be the latest one (so it gets kept in the history longer)

Reply:
![image](https://user-images.githubusercontent.com/10514821/67512134-d75e6a00-f6a0-11e9-910f-b2740e8ccba7.png)
@juliusv Everything seems to be fine. Did you refresh the page after setting the items in the storage?

Reply:
So the first issue is that local storage seems unbounded:

- Delete `history` and `enable-query-history` local storage items via Chrome's inspector (Application->Local Storage)
- Enter >50 different queries.
- Watch the local storage entry grow to >50 entries. This should be limited to not grow infinitely.

The other issue with suggestions now showing up I get with:

- Delete both local storage items again, enable query history.
- Enter and execute `1` ... `10` as 10 different queries.
- Enter `1` and get no autocomplete results (neither `1` nor `10`). The numbers `2` ... `9` do give me autocomplete results in this case btw.
- When I reload the page, then I also get `1` and `10` as autocomplete results, but that shouldn't be needed.